### ISSUE TITLE
Parsing an integer value when JSON contains a decimal point

### ISSUE THREAD

**Issue by giedriusbruzas** [mausch/Fleece#17]
let result : int ParseResult = parseJSON "1.1"
    let (Choice1Of2 value) = result
    //val value : int = 1

I feel this is incorrect and should instead result in a failure.
There's possibly other similar edge cases like this one.

**Comment by wallymathieu** [comment]
This also causes trouble when trying to do lensing.

**Comment by wallymathieu** [comment]
`JNumber s` implies that `s` is JsonPrimitive ...

**Comment by wallymathieu** [comment]
Seems like F# Data us fixing this as well: https://www.nuget.org/packages/FSharp.Data/3.0.0-beta4 :
> (Breaking Change) Don't silently convert decimals and floats to integers in JsonProvider.

**Comment by wallymathieu** [comment]
https://github.com/fsharp/FSharp.Data/pull/1185

**Comment by gusty** [comment]
I think this behavior is not correct at all.
Anyway, I also think this should be corrected in the original library (as FSharp.Data just did).
We should forward this issue to Newtonsoft and System.Json
Also we should note that Fleece is a Json mapper for different libraries, since different libraries have different behavior (there is no agreed standard for Json <-> .NET) it's not up to Fleece to try to fix them.

**Comment by wallymathieu** [comment]
I agree. It only makes sense to align with the libraries. Right now JNumber active pattern has some interesting behavior.

**Comment by gusty** [comment]
I don't know if this is related:

    1 |> toJson |> string;;
    val it : string = "1"

but

    (2,1) |> toJson |> string;;
    val it : string = "[
      2.0,
      1.0
    ]"

**Comment by wallymathieu** [comment]
[Fleece.fs#L117-L118](https://github.com/mausch/Fleece/blob/fb83c00a92c5af1566d9d5a31a8fed6ffd32142a/Fleece/Fleece.fs#L117-L118) (Newtonsoft.Json):
```f#
        | JTokenType.Integer -> JNumber  o
        | JTokenType.Float   -> JNumber  o
```
[Fleece.fs#L223-L224](https://github.com/mausch/Fleece/blob/fb83c00a92c5af1566d9d5a31a8fed6ffd32142a/Fleece/Fleece.fs#L223-L224) (FSharp.Data):
```f#
        | JsonValue.Number _ as x -> JNumber x
        | JsonValue.Float _ as x  -> JNumber x
```
while for System Json there doesn't seem to be such a distinction

**Comment by wallymathieu** [comment]
@gusty that's probably a bug in Newtonsoft.Json.

**Comment by wallymathieu** [comment]
We also see this kind of behavior: https://github.com/JamesNK/Newtonsoft.Json/issues/1400
So I guess it's a tradeoff what Json parser you use.

**Comment by wallymathieu** [comment]
After looking at the behavior a bit more closely, it could be something I wrote when making the Newtonsoft.Json fleece port.

**Comment by wallymathieu** [comment]
I've added a PR that documents the behavior. Note that Fsharp.Data works the way you want.
